include/clad/Differentiator/NumericalDiff.h

```diff
@@ -0,0 +1,418 @@
+  ///   }
+  ///   return arg;
+  /// }
+  /// \endcode
+  ///
+  /// \param[in] arg The argument to update.
+  /// \param[in] idx The index of the current parameter in the target function
+  /// parameter pack.
+  /// \param[in] currIdx The index value of the parameter to be
+  /// updated.
+  /// \param[in] multiplier A multiplier to compound with h before
+  /// updating 'arg'.
+  /// \param[out] h_val The h value that was set for each
+  /// argument.
+  /// \param[in] bufferManager A buffer manager to request for buffer space
+  /// while forwarding reference/pointer args to the target function.
+  /// \param[in] n The size of the input pointer/array.
+  /// \param[in] i The specific index to get the adjusted h for.
+  ///
+  /// \returns The updated argument.
+  template <typename T, typename U>
+  T* updateIndexParamValue(T* arg, std::size_t idx, std::size_t currIdx,
+                           int multiplier, precision& h_val,
+                           ManageBufferSpace<U>& bufferManager,
+                           std::size_t n = 0, std::size_t i = 0) {
+    // malloc some buffer space that we will free later.
+    // this is required to make sure that we are retuning a deep copy
+    // that is valid throughout the scope of the central_diff function.
+    // USER FUNCTION SHOULD NOT DELETE. IT WILL LEAD TO UB.
+    T* temp = bufferManager.template make_buffer_space<T>(n);
+    // deepcopy
+    for (std::size_t j = 0; j < n; j++) {
+      temp[j] = arg[j];
+    }
+    if (idx == currIdx) {
+      h_val = get_h(temp[i]);
+      // update the specific array value.
+      temp[i] += h_val * multiplier;
+    }
+    return temp;
+  }
+
+  /// A helper function to calculate the numerical derivatve of a target
+  /// function.
+  ///
+  /// \param[in] f The target function to numerically differentiate.
+  /// \param[out] _grad The gradient array reference to which the gradients will
+  /// be written.
+  /// \param[in] idxSeq The index sequence associated with the input
+  /// parameter pack.
+  /// \param[in] args The arguments to the function to differentiate.
+  template <typename F, std::size_t... Ints,
+            typename RetType = typename clad::return_type<F>::type,
+            typename... Args>
+  void central_difference_helper(
+      F f, clad::tape_impl<clad::array_ref<RetType>>& _grad,
+      clad::IndexSequence<Ints...> idxSeq, Args&&... args) {
+
+    std::size_t argLen = sizeof...(Args);
+    ManageBufferSpace<void*> bufferManager;
+    // loop over all the args, selecting each arg to get the derivative with
+    // respect to.
+    for (std::size_t i = 0; i < argLen; i++) {
+      precision h1, h2, h;
+      std::size_t argVecLen = _grad[i].size();
+      for (std::size_t j = 0; j < argVecLen; j++) {
+        // calculate f[x+h, x-h]
+        // f(..., x+h,...)
+        precision xaf = f(updateIndexParamValue(std::forward<Args>(args), Ints,
+                                                i, 1, h1, bufferManager,
```

Comment:
```suggestion
                                                i, /*multiplier=*/1, h1, bufferManager,
```
please update all such uses.